Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -1636,22 +1673,126 @@ void CGContextSetCMYKFillColor(CGContextRef context, CGFloat cyan, CGFloat magen
+
+    CGAffineTransform transform = __BitmapBrushTransformation(context,
+                                                              { { 0, 0 }, size.width, size.height },
+                                                              d2dBitmap->GetPixelSize(),
+                                                              _CGPatternGetTransformation(pattern));
+
+    ComPtr<ID2D1BitmapBrush1> bitmapBrush;
+    ComPtr<ID2D1DeviceContext> deviceContext = context->DeviceContext();
+    RETURN_IF_FAILED(deviceContext->CreateBitmapBrush(d2dBitmap.Get(),
+                                                      D2D1::BitmapBrushProperties1(D2D1_EXTEND_MODE_WRAP,
+                                                                                   D2D1_EXTEND_MODE_WRAP,
+                                                                                   context->CurrentGState().GetInterpolationModeForCGImage(
+                                                                                       tileImage.get())),
+                                                      D2D1::BrushProperties(1 /* the global alpha will be set during the draw sequence*/,
+                                                                            __CGAffineTransformToD2D_F(transform)),
+                                                      &bitmapBrush));
+
+    *brush = bitmapBrush.Detach();
+    return S_OK;
+}
+
 /**
- @Status Stub
- @Notes
+ @Status Caveat
+ @Notes only supports RGBA components (due to CGContextSetFillColor only supporting RGBA)
 */
-void CGContextSetStrokePattern(CGContextRef context, CGPatternRef pattern, const CGFloat* components) {
+void CGContextSetFillPattern(CGContextRef context, CGPatternRef pattern, const CGFloat* components) {
     NOISY_RETURN_IF_NULL(context);
-    UNIMPLEMENTED();
+    NOISY_RETURN_IF_NULL(pattern);
+    NOISY_RETURN_IF_NULL(components);
+
+    ComPtr<ID2D1BitmapBrush1> bitmapBrush;
+    FAIL_FAST_IF_FAILED(_CreatePatternBrush(context, pattern, components, &bitmapBrush, [&](CGContextRef drawingContext) {
+        CGContextSetFillColorSpace(drawingContext, context->FillColorSpace());
+        CGContextSetFillColor(drawingContext, components);
+        return S_OK;
+    }));
+    // set the fill brush
+    FAIL_FAST_IF_FAILED(bitmapBrush.As(&context->CurrentGState().fillBrush));
 }
 
 /**
- @Status Interoperable
+ @Status Caveat
+ @Notes only supports RGBA components (due to CGContextSetFillColor only supporting RGBA)
 */
-void CGContextSetFillPattern(CGContextRef context, CGPatternRef pattern, const CGFloat* components) {
+void CGContextSetStrokePattern(CGContextRef context, CGPatternRef pattern, const CGFloat* components) {
     NOISY_RETURN_IF_NULL(context);
-    UNIMPLEMENTED();
+    NOISY_RETURN_IF_NULL(pattern);
+    NOISY_RETURN_IF_NULL(components);
+
+    ComPtr<ID2D1BitmapBrush1> bitmapBrush;
+    FAIL_FAST_IF_FAILED(_CreatePatternBrush(context, pattern, components, &bitmapBrush, [&](CGContextRef drawingContext) {
```

Comment:
not the same

---
In reply to: [93077490](https://github.com/Microsoft/WinObjC/pull/1586#discussion_r93077490) [](ancestors = 93077490)